File: aprs_kalman.m
%% APRS_kalman.m
% Loads an aprs.fi export and predicts next state as xy coordinates
close all;
%% Load Data
% Most of these variables are important; chiefly x and y, but also
% speed/time/altitude
datacsv = readtable('datasets/w2kgy-12_balloon.csv');
% let's start with just altitude - more linear
x_data = table2array(datacsv(:,"lng"));
y_data = table2array(datacsv(:,"lat"));
alt_data = table2array(datacsv(:,"altitude"));
alt_data = alt_data/1e3; %convert to kft
vel_data = table2array(datacsv(:,"speed"));
time_data = table2array(datacsv(:,"lasttime"));
% plot(alt_data) % this actually doesn't need much sanitization
data_cutoff = 51; %how many to use as training data
alt_train = alt_data(1:data_cutoff);
alt_est = alt_data(1+data_cutoff:end);
time_train = time_data(1:data_cutoff);
time_est = time_data(1+data_cutoff:end);

%% Estimate
% Calculate Kalman inputs A,B,Rn,H,Rv,x
time_diff = zeros(1,length(time_train)-1);
for k=1:length(time_diff)
    time_diff(k) = second(time_train(k)) -second(time_train(k+1));
end
T = mean(time_diff);
A = [1 T; 0 1]; B = [T^2/2 T];
var_alt = var(alt_train);
%% Predict and Format

% Generate vector process y(n)
N = 50;
% Generate the observation process x(n)
g1 = 0.25;
xn = alt_train;
x_std = std(xn);
% Design KF parameters
H = [1 0]; G = 0.25;
Rv = 0.25^2; R_eta = eye(1);
% Initialization
y_post = [0;1]; R_post = zeros(2);
IRv = eye(size(Rv)); IR = eye(size(R_post));
y_hat = zeros(N+1,1); gain = zeros(N+1,2); mse = y_hat;
% Tracking
for n = 0:N
R_pri = A*R_post*A' + B*R_eta*B';
y_pri = A*y_post;
x_pri = H*y_pri;
Rw = H*R_pri*H'+Rv;
K = R_pri*H'*(Rw\IRv);
y_post = y_pri + K*(xn(n+1) - x_pri);
R_post = (IR-K*H)*R_pri;
y_hat(n+1) = y_post(2);
gain(n+1,:) = K';
mse(n+1) = R_post(1,1);
end
y_hat = [y_hat(2:N+1);y_hat(N+1)];

n = 0:N; figure;
plot(n,xn,':',n,y_hat,'g--','linewidth',1);
ylabel('Amplitude'); xlabel('Index n');
legend('x(n)','yhat(n)'); axis([0,N,0,max(xn)+10]);
set(gca,'xtick',(0:20:N),'ytick',(0:5:max(xn)+10)); grid;
title('Estimation of Altitude Using Kalman Filter');

File: aprs_kalman.py
import matplotlib.pyplot as plt
import matplotlib.animation as animation
import numpy as np

# Load Data

alt_data = np.loadtxt('datasets/w2kgy-12_balloon.csv', delimiter=',', skiprows=1,usecols=6)
time_data = np.loadtxt('datasets/w2kgy-12_balloon.csv', delimiter=',', skiprows=1,usecols=0, dtype=np.datetime64)

alt_data = alt_data*0.3048/1000 #convert to km
data_cutoff = 31
alt_train = alt_data[0:data_cutoff]
alt_est = alt_data[1+data_cutoff:]

t_diff = np.zeros(len(time_data))
for t in range(0,len(time_data)-1):
    t_diff[t] = time_data[t+1]-time_data[t]
T = np.mean(t_diff)

A = np.array([[1, T],[0, 1]])
B = np.array([[0.5*(T**2)],[T]])
N = data_cutoff-1
xn = alt_train-np.min(alt_train)
x_std = np.cov(xn)**2
# KF Parameters
H = np.array([[0,1]])
Rv = np.cov(xn)**2
R_eta = np.eye(1)
# Initialization
y_post = np.array([[0],[1]])
R_post = np.zeros((2,2))

IR = np.eye(np.size(R_post[:,0]))
y_hat = np.zeros(N+1)

for n in range(0,N):
    # Predict
    R_pri = A @ R_post @ A.T + B @ R_eta @ B.T
    y_pri = A @ y_post
    x_pri = H @ y_pri
    # Update
    Rw = H @R_pri @ H.T + Rv
    K = R_pri @ H.T @ np.linalg.inv(Rw)
    y_post = y_pri + K @ (xn[n+1]-x_pri)
    R_post = (IR-K @ H) @ R_pri
    y_hat[n+1] = y_post[1]

#y_hat = [y_hat[1:-1],y_hat[-1]]
# xn = xn[0:-2]
n = np.arange(0,N+1)
xn = np.roll(xn,1)
xn[0] = 0
nx = np.append(0,n)
nx = nx[0:-1]
# plot steps, -> xn and y_hat
fig, ax = plt.subplots()  # Create a figure containing a single axes.
meas_alt = ax.plot(nx[0], xn[0], label='Measured Altitude', color='blue', marker=".",linewidth=2, linestyle='--')[0]  # Plot some data on the axes.
est_alt = ax.plot(n[0],y_hat[0], label='Estimated Altitude', color='orange', marker=".",linestyle=':')[0]
ax.set(xlim=[0, N+1], ylim=[0, np.max(xn)], xlabel='Measurement Step', ylabel='Altitude (km)')
ax.set_title('1D Altitude Kalman Filter')
ax.grid(True)
ax.legend()

def update(frame):
    # update the line plot:
    meas_alt.set_xdata(nx[:frame])
    meas_alt.set_ydata(xn[:frame])
    est_alt.set_xdata(n[:frame])
    est_alt.set_ydata(y_hat[:frame])
    return (meas_alt, est_alt)


ani = animation.FuncAnimation(fig=fig, func=update, frames=N+1, interval=750)
#ani.to_html5_video(64e3)
ani.save('Kalman1D.mp4')
plt.show()
